### PR TITLE
Convert crm_ticket to use formatted output

### PR DESCRIPTION
**Pull request by clumens**
This also includes the patch from #3356, but I assume I'll have to rebase this a couple times so I'm not worried about that.

One significant thing to note is that I don't think the new ticket schema is complete.  There's a block in the ticket_xml function that adds arbitrary key/value pairs as attributes to the ticket's XML node.  I'm not sure what all could be in those pairs, so at the moment I've just left it out of the schema.  I'll have to add something, though, and if they can be potentially anything, the schema is a little less useful.

### CODE REVIEW

File: tools/crm_ticket.c
Comment (kgaillot) [inline tools/crm_ticket.c:361]: This is only valid for text format, need to separate from default, or maybe just use info()

**Review by kgaillot — COMMENTED** [review]
yay

File: cts/cli/regression.tools.exp
Comment (clumens) [reply cts/cli/regression.tools.exp:2953]: It looks grammatically possible to reuse just the rsc_ticket portion of the constraints schema here, but the reason it won't work is that in that schema, rsc is an IDREF.  This needs to refer to a defined resource with an ID somewhere, and the ticket XML isn't going to have any resources.

File: tools/crm_ticket.c
Comment (clumens) [reply tools/crm_ticket.c:361]: Changed to use info().

File: cts/cli/regression.tools.exp
Comment (clumens) [reply cts/cli/regression.tools.exp:2329]: Fixed.

File: tools/crm_ticket.c
Comment (clumens) [reply tools/crm_ticket.c:759]: No, there isn't.  Removing this will change the list output to this style:
```
<list name="tickets" count="2">
   <ticket ...>
   <ticket ...>
</list>
```

I no longer remember why we do one over the other.  The non-simple list style seems less XMLish to me anymore, but we've got a mix of programs using each way.

File: cts/cli/regression.tools.exp
Comment (kgaillot) [reply cts/cli/regression.tools.exp:2953]: What do you think of something like:

    <resources>
        <rsc id="..."/>
    </resources>
    <tickets>
        <rsc_ticket .../>
    </tickets>

Either that or just keep a separate rsc_ticket schema in api.

File: tools/crm_ticket.c
Comment (kgaillot) [reply tools/crm_ticket.c:759]: IIRC we added simple list to maintain previous behavior somewhere, and I guess it just sort of grew from there. I don't have a preference one way or the other as to the style, but I suspect we have the problem where the libpacemaker API output won't follow the schema used by the tools. Maybe we should convert it to a begin_list() argument

File: cts/cli/regression.tools.exp
Comment (clumens) [reply cts/cli/regression.tools.exp:2953]: I like adding the resources to the schema.  I think you're going to enjoy the ridiculous stuff I've had to do to reuse the constraint schema.

File: tools/crm_ticket.c
Comment (clumens) [reply tools/crm_ticket.c:759]: It looks like this (and --xml-legacy) were added to preserve the --as-xml behavior in crm_mon.  It is also currently in use in pacemakerd, crmadmin, crm_node, crm_simulate, and crm_resource.  We probably shouldn't have used it in those other tools, and I assume it's too late to start taking it back out.

I also think you're right that libpacemaker won't follow the same schema because of this - I don't see us forcing the argument anywhere in the library.  It's definitely time to start adding further tests here.

I guess I'm not quite sure how to start cleaning up this mess.

Comment (clumens) [reply tools/crm_ticket.c:759]: But I think it's worth revisiting one of my previous comments - the new <list> lists don't really strike me as especially XMLish.  The simple list output looks more like what I would expect if I were just looking at a random XML blob.  I don't remember why we thought that was a good idea.

Comment (kgaillot) [reply tools/crm_ticket.c:759]: > It looks like this (and --xml-legacy) were added to preserve the --as-xml behavior in crm_mon. It is also currently in use in pacemakerd, crmadmin, crm_node, crm_simulate, and crm_resource. We probably shouldn't have used it in those other tools, and I assume it's too late to start taking it back out.

Yeah ...
 
> I also think you're right that libpacemaker won't follow the same schema because of this - I don't see us forcing the argument anywhere in the library. It's definitely time to start adding further tests here.
> 
> I guess I'm not quite sure how to start cleaning up this mess.

We should probably drop the command-line option. We could make it a new global, or a begin_list() argument, or maybe something like substitutions[]

> But I think it's worth revisiting one of my previous comments - the new lists don't really strike me as especially XMLish. The simple list output looks more like what I would expect if I were just looking at a random XML blob. I don't remember why we thought that was a good idea.

I don't have a preference. I believe the original idea was that automated processing of the XML would be easier if it reused schema elements as much as possible.

**Comment by clumens** [comment]
Pushed everything except for the list XML changes, which I'm still thinking about.

File: tools/crm_ticket.c
Comment (kgaillot) [inline tools/crm_ticket.c:417]: rsc_ticket now

**Review by kgaillot — COMMENTED** [review]
looks good so far

File: tools/crm_ticket.c
Comment (clumens) [reply tools/crm_ticket.c:780]: In trying to decide what to do here, I've noticed that we basically do three different things:

```
<list name="things" count="3">
   <item name="thing" />
   <item name="thing" />
   <item name="thing" />
</list>
```

I think only stonith_admin does that.

```
<things>
   <thing />
   <thing />
   <thing />
</things>
```

Lots of places do this.

```
<pacemaker-result>
   <thing />
   <thing />
   <thing />
   <status code="" message="" />
</pacemaker-result>
```

A surprisingly large number of places do this.

Comment (clumens) [reply tools/crm_ticket.c:780]: Rebasing on #3365 takes care of this.

File: xml/api/crm_mon-2.33.rng
Comment (nrwahl2) [inline xml/api/crm_mon-2.33.rng:1]: This should be `crm_mon-2.34.rng`. The main branch is already at API version 2.34.

File: xml/api/ticket-2.33.rng
Comment (nrwahl2) [inline xml/api/ticket-2.33.rng:6]: I don't think `combine="choice"` is necessary. The original `start` component of the `constraints-3.9.rng` schema will be replaced by this one completely, so we end up with only one `start` element. (i.e., it doesn't get combined with anything)

> Define the components of an element to be the children of the element together with the components of any `div` child elements. If the `include` element has a `start` component, then the `grammar` element must have a `start` component. If the `include` element has a `start` component, then all `start` components are removed from the `grammar` element. If the `include` element has a `define` component, then the `grammar` element must have a `define` component with the same name. For every `define` component of the `include` element, all `define` components with the same name are removed from the `grammar` element.

https://relaxng.org/spec-20011203.html#IDAG3YR

File: xml/api/crm_mon-2.33.rng
Comment (nrwahl2) [reply xml/api/crm_mon-2.33.rng:1]: Keep an eye on this before merging. #3338 also introduces an API version 2.34, so whichever gets merged second will need to be 2.35.

**Comment by clumens** [comment]
Rebased against the recently merged simple list stuff.

File: tools/crm_ticket.c
Comment (kgaillot) [inline tools/crm_ticket.c:464]: If backward compatibility weren't an issue, it would be nice to implement friendly text output

File: tools/crm_ticket.c
Comment (kgaillot) [inline tools/crm_ticket.c:448]: Would it make sense as a general principle to pop back up to the original level at the end of a custom message?

File: lib/pengine/pe_output.c
Comment (kgaillot) [inline lib/pengine/pe_output.c:3170]: Does booth create expires? Pacemaker doesn't seem to use it. Might as well create a constant for it while we're here ...

File: lib/pengine/pe_output.c
Comment (kgaillot) [inline lib/pengine/pe_output.c:3261]: space after while

File: lib/pengine/pe_output.c
Comment (kgaillot) [inline lib/pengine/pe_output.c:3255]: xml can always add details (unless we need it not to somewhere for backward compatibility)

**Review by kgaillot — COMMENTED** [review]
awesome

File: lib/pengine/pe_output.c
Comment (clumens) [reply lib/pengine/pe_output.c:3255]: Fixed.  The only catch here was making sure we don't also output a `granted=` attribute, since that duplicates the status attribute.  While I don't think that causes any problems, nothing was outputting it previously.

File: tools/crm_ticket.c
Comment (clumens) [reply tools/crm_ticket.c:448]: Hm, yes definitely.  At the moment, I think we are okay here but it's sloppy and in the future if we go to reuse this code, anything that generates XML afterwards is going to be adding elements below unrelated parent elements.

**Comment by fabbione** [comment]
retest this please

**Comment by clumens** [comment]
Hold off on this for a bit - I'm writing some tests for a PR that builds on top of this one, and I think those tests are going to cause me to have to make some changes here.

File: xml/api/ticket-2.34.rng
Comment (clumens) [inline xml/api/ticket-2.34.rng:58]: @kgaillot Looking at the XML files in cts/scheduler/xml, it looks like there's only ever `<ticket_state>` elements, and that those elements contain id, granted, last-granted, standby, etc.  And then looking at `modify_ticket_state` in crm_ticket.c, it looks like any of the attributes that could be added to a ticket would also go into the `<ticket_state>`.

So with that in mind, I feel like perhaps the schema here is redundant.  The fact that crm_mon is already using a ticket schema makes it a little more complicated, but maybe the right thing to do is simply remove the `<ticket_state>` portion.  I think everything in there is covered by the higher up element.  I think I'd need to move the generic attribute stuff up a level, too.

One thing that's confusing me a little bit here is that in the CIB, it's always `<ticket_state>`, but in the API, it would be `<ticket>`.  Maybe that's only confusing on the implementation side, and it wouldn't be a problem for users.

Comment (kgaillot) [reply xml/api/ticket-2.34.rng:58]: Sounds good. The CIB has both `rsc_ticket` in constraints and `ticket_state` in state. I don't think it's a problem that the API uses something different to present the info to the user.

**Comment by clumens** [comment]
Rebased, updated the schema as previously discussed.

**Review by kgaillot — APPROVED** [review]
looks good to me, let me know when you're ready to merge

**Comment by clumens** [comment]
I've got another PR in the works that builds on this one and moves crm_ticket stuff out into a library.  At the moment, all I've done is constraints.  That PR is also a useful way of verifying that the schema here is correct.  So, you can merge this one now and accept the possibility of future schema changes (before the next release, I'm sure) or we can sit on this one a while longer until I'm completely done with the next PR.  I don't forsee a need for any schema changes, but of course I could be wrong about that.

I can also push that one as a WIP if it helps you figure out what to do.

**Comment by kgaillot** [comment]
Seems reasonable enough

**Pull request merged**
